File: packages/in_app_purchase/in_app_purchase_storekit/example/shared/RunnerTests/FIAPPaymentQueueDeleteTests.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import <XCTest/XCTest.h>
#import "FIAObjectTranslator.h"
#import "FIAPaymentQueueHandler.h"
#import "Stubs.h"

@import in_app_purchase_storekit;

API_AVAILABLE(ios(13.0))
API_UNAVAILABLE(tvos, macos, watchos)
@interface FIAPPaymentQueueDelegateTests : XCTestCase

@property(nonatomic, strong) SKPaymentTransaction *transaction;
@property(nonatomic, strong) SKStorefront *storefront;

@end

@implementation FIAPPaymentQueueDelegateTests

- (void)setUp {
  NSDictionary *transactionMap = @{
    @"transactionIdentifier" : [NSNull null],
    @"transactionState" : @(SKPaymentTransactionStatePurchasing),
    @"payment" : [NSNull null],
    @"error" : [FIAObjectTranslator getMapFromNSError:[NSError errorWithDomain:@"test_stub"
                                                                          code:123
                                                                      userInfo:@{}]],
    @"transactionTimeStamp" : @([NSDate date].timeIntervalSince1970),
    @"originalTransaction" : [NSNull null],
  };
  self.transaction = [[SKPaymentTransactionStub alloc] initWithMap:transactionMap];

  NSDictionary *storefrontMap = @{
    @"countryCode" : @"USA",
    @"identifier" : @"unique_identifier",
  };
  self.storefront = [[SKStorefrontStub alloc] initWithMap:storefrontMap];
}

- (void)tearDown {
}

- (void)testShouldContinueTransaction {
  if (@available(iOS 13.0, *)) {
    MethodChannelStub *channelStub = [[MethodChannelStub alloc] init];
    channelStub.invokeMethodChannelWithResultsStub =
        ^(NSString *_Nonnull method, id _Nonnull arguments, FlutterResult _Nullable result) {
          XCTAssertEqualObjects(method, @"shouldContinueTransaction");
          XCTAssertEqualObjects(arguments,
                                [FIAObjectTranslator getMapFromSKStorefront:self.storefront
                                                    andSKPaymentTransaction:self.transaction]);
          result(@NO);
        };

    FIAPPaymentQueueDelegate *delegate =
        [[FIAPPaymentQueueDelegate alloc] initWithMethodChannel:channelStub];

    BOOL shouldContinue = [delegate paymentQueue:[[SKPaymentQueueStub alloc] init]
                       shouldContinueTransaction:self.transaction
                                    inStorefront:self.storefront];

    XCTAssertFalse(shouldContinue);
  }
}

- (void)testShouldContinueTransaction_should_default_to_yes {
  if (@available(iOS 13.0, *)) {
    MethodChannelStub *channelStub = [[MethodChannelStub alloc] init];
    FIAPPaymentQueueDelegate *delegate =
        [[FIAPPaymentQueueDelegate alloc] initWithMethodChannel:channelStub];

    channelStub.invokeMethodChannelWithResultsStub =
        ^(NSString *_Nonnull method, id _Nonnull arguments, FlutterResult _Nullable result) {
          XCTAssertEqualObjects(method, @"shouldContinueTransaction");
          XCTAssertEqualObjects(arguments,
                                [FIAObjectTranslator getMapFromSKStorefront:self.storefront
                                                    andSKPaymentTransaction:self.transaction]);
        };

    BOOL shouldContinue = [delegate paymentQueue:[[SKPaymentQueueStub alloc] init]
                       shouldContinueTransaction:self.transaction
                                    inStorefront:self.storefront];

    XCTAssertTrue(shouldContinue);
  }
}

#if TARGET_OS_IOS
- (void)testShouldShowPriceConsentIfNeeded {
  if (@available(iOS 13.4, *)) {
    MethodChannelStub *channelStub = [[MethodChannelStub alloc] init];
    FIAPPaymentQueueDelegate *delegate =
        [[FIAPPaymentQueueDelegate alloc] initWithMethodChannel:channelStub];

    channelStub.invokeMethodChannelWithResultsStub =
        ^(NSString *_Nonnull method, id _Nonnull arguments, FlutterResult _Nullable result) {
          XCTAssertEqualObjects(method, @"shouldShowPriceConsent");
          XCTAssertNil(arguments);
          result(@NO);
        };

    BOOL shouldShow =
        [delegate paymentQueueShouldShowPriceConsent:[[SKPaymentQueueStub alloc] init]];

    XCTAssertFalse(shouldShow);
  }
}
#endif

#if TARGET_OS_IOS
- (void)testShouldShowPriceConsentIfNeeded_should_default_to_yes {
  if (@available(iOS 13.4, *)) {
    MethodChannelStub *channelStub = [[MethodChannelStub alloc] init];
    FIAPPaymentQueueDelegate *delegate =
        [[FIAPPaymentQueueDelegate alloc] initWithMethodChannel:channelStub];

    channelStub.invokeMethodChannelWithResultsStub =
        ^(NSString *_Nonnull method, id _Nonnull arguments, FlutterResult _Nullable result) {
          XCTAssertEqualObjects(method, @"shouldShowPriceConsent");
          XCTAssertNil(arguments);
        };

    BOOL shouldShow =
        [delegate paymentQueueShouldShowPriceConsent:[[SKPaymentQueueStub alloc] init]];

    XCTAssertTrue(shouldShow);
  }
}
#endif

@end

File: packages/in_app_purchase/in_app_purchase_storekit/example/shared/RunnerTests/FIAPPaymentQueueDeleteTests.swift
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation
import StoreKit
import XCTest

@testable import in_app_purchase_storekit

@available(iOS 13.0, *)
final class FIAPPaymentQueueDelegateTests: XCTestCase {
  var transaction: SKPaymentTransaction!
  var storefront: SKStorefront!

  override func setUp() {
    super.setUp()
    let transactionMap: [String: Any] = [
      "transactionIdentifier": NSNull(),
      "transactionState": SKPaymentTransactionState.purchasing.rawValue,
      "payment": NSNull(),
      "error": FIAObjectTranslator.getMapFrom(
        NSError(domain: "test_stub", code: 123, userInfo: [:])),
      "transactionTimeStamp": Date().timeIntervalSince1970,
      "originalTransaction": NSNull(),
    ]
    self.transaction = SKPaymentTransactionStub(map: transactionMap)

    let storefrontMap: [String: Any] = [
      "countryCode": "USA",
      "identifier": "unique_identifier",
    ]
    self.storefront = SKStorefrontStub(map: storefrontMap)
  }

  override func tearDown() {
    super.tearDown()
  }

  func testShouldContinueTransaction() {
    let channelStub = MethodChannelStub()
    channelStub.invokeMethodChannelWithResultsStub = { method, arguments, result in
      XCTAssertEqual(method, "shouldContinueTransaction")
      XCTAssertEqual(
        arguments as! NSDictionary,
        FIAObjectTranslator.getMapFrom(
          self.storefront,
          andSKPaymentTransaction: self.transaction) as NSDictionary)

      guard let result = result else {
        XCTFail("Result should not be nil")
        return
      }
      result(false)
    }

    let delegate = FIAPPaymentQueueDelegate(methodChannel: channelStub)

    let shouldContinue = delegate.paymentQueue(
      SKPaymentQueueStub(),
      shouldContinue: self.transaction,
      in: self.storefront)

    XCTAssertFalse(shouldContinue)
  }

  func testShouldContinueTransactionShouldDefaultToYes() {
    let channelStub = MethodChannelStub()
    let delegate = FIAPPaymentQueueDelegate(methodChannel: channelStub)

    channelStub.invokeMethodChannelWithResultsStub = { method, arguments, result in
      XCTAssertEqual(method, "shouldContinueTransaction")
      XCTAssertEqual(
        arguments as! NSDictionary,
        FIAObjectTranslator.getMapFrom(
          self.storefront,
          andSKPaymentTransaction: self.transaction) as NSDictionary)
    }

    let shouldContinue = delegate.paymentQueue(
      SKPaymentQueueStub(),
      shouldContinue: self.transaction,
      in: self.storefront)

    XCTAssertTrue(shouldContinue)
  }

  #if TARGET_OS_IOS
    func testShouldShowPriceConsentIfNeeded() throws {
      if #available(iOS 13.4, *) {
        let channelStub = MethodChannelStub()
        let delegate = FIAPPaymentQueueDelegate(methodChannel: channelStub)

        channelStub.invokeMethodChannelWithResultsStub = { method, arguments, result in
          XCTAssertEqual(method, "shouldShowPriceConsent")
          XCTAssertNil(arguments)

          guard let result = result else {
            XCTFail("Result should not be nil")
            return
          }
          result(false)
        }

        let shouldShow = delegate.paymentQueueShouldShowPriceConsent(SKPaymentQueueStub())

        XCTAssertFalse(shouldShow)
      }
    }

    func testShouldShowPriceConsentIfNeededShouldDefaultToYes() {
      if #available(iOS 13.4, *) {
        let channelStub = MethodChannelStub()
        let delegate = FIAPPaymentQueueDelegate(methodChannel: channelStub)

        channelStub.invokeMethodChannelWithResultsStub = { method, arguments, result in
          XCTAssertEqual(method, "shouldShowPriceConsent")
          XCTAssertNil(arguments)
        }

        let shouldShow = delegate.paymentQueueShouldShowPriceConsent(SKPaymentQueueStub())

        XCTAssertTrue(shouldShow)
      }
    }
  #endif
}
